### DISCUSSION TITLE
[plugins] Electron- or Node-specific code

### DISCUSSION THREAD

**Discussion by thegecko** [eclipse-theia/theia#7826] [General]
Hi all,

I notice the plugin system doesn't support the Electron-specific layers in the same way Theia packages do:

https://github.com/eclipse-theia/theia/blob/master/packages/plugin-ext/src/common/plugin-protocol.ts#L46

What mechanisms do people use to add Electron or Node specific funtionality in plugins and/or should we add support for splitting this in plugins? e.g.

```json
  "theiaPlugin": {
    "backend": "lib/backend-node",
    "backendElectron": "lib/backend-electron"
  },
```

**Comment by akosyakov** [comment]
VS Code extensions runs in pure Node.js process, no? I wonder what concrete use case you had in mind.

**Reply by thegecko** [reply]
I need a single plugin to act differently depending on the host :)

**Reply by akosyakov** [reply]
Maybe you need: https://code.visualstudio.com/api/references/vscode-api#UIKind Not sure how well it is supported right now. If not please file an issue to fix it.

I really would prefer that we don't deviate from VS Code in regards to plugins and when it is possible build stuff based on VS Code machinery, not invent ours. It is very hard to catch up.
